utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2073 @@
+    return result;
+  }
+
+  assert (opt_db_);
+  Transaction* txn = static_cast<OptimisticTransactionDB*>(opt_db_.get())->BeginTransaction(write_options_);
+  OptimisticTransactionImpl *otxn = dynamic_cast<OptimisticTransactionImpl*>(txn);
+  assert (otxn != nullptr);
+
+  bool no_relocation_ttl = (has_ttl && tt > bfptr->GetTTLRange().second);
+  bool no_relocation_lsmdel = (bfptr->GetFileSize() ==
+    (bfptr->deleted_size_ + blob_log::BlobLogFooter::kFooterSize));
+
+  bool no_relocation = no_relocation_ttl || no_relocation_lsmdel;
+
+  std::shared_ptr<BlobFile> newfile;
+  std::shared_ptr<blob_log::Writer> new_writer;
+
+  if (!no_relocation) {
+    std::string reason("GC of ");
+    reason += bfptr->PathName();
+    newfile = newBlobFile(reason);
+
+    new_writer = checkOrCreateWriter_locked(newfile);
+    newfile->setHeader_locked(header);
+    newfile->file_size_ = blob_log::BlobLogHeader::kHeaderSize;
+    s = new_writer->WriteHeader(newfile->header_);
+
+    if (!s.ok()) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "File: %s - header writing failed", newfile->PathName().c_str());
+      no_relocation = true;
+      // should be a much severe error
+    }
+
+    // read the blob because you have to write it back to
+    // new file
+    shallow = blob_log::Reader::READ_LEVEL_HDR_FOOTER_KEY_BLOB;
+  }
+
+  uint32_t num_deletes = 0;
+
+  while (reader->ReadRecord(&record, shallow).ok()) {
+    ++blob_count;
+
+    // this particular TTL has expired
+    if (no_relocation_ttl || (has_ttl && tt > record.GetTTL())) {
+      txn->Delete(cfh, record.Key());
+      num_deletes++;
+      continue;
+    }
+
+    SequenceNumber seq = kMaxSequenceNumber;
+    bool found_record_for_key = false;
+    s = db_impl_->GetLatestSequenceForKey(sv, record.Key(), false,
+      &seq, &found_record_for_key);
+
+    bool del_this = s.ok() && (!found_record_for_key || seq !=
+      record.GetSN());
+
+    if (del_this) {
+      // stil could have a TOCTOU
+      txn->Delete(cfh, record.Key());
+      num_deletes++;
+      continue;
+    }
+
+    {
+      std::string index_entry;
+      PutVarint64(&index_entry, newfile->BlobFileNumber());
+
+      BlockHandle handle;
+      auto raw_block_size = record.Blob().size();
+      handle.set_size(raw_block_size);
+
+      // these are returned values
+      uint64_t blob_offset = 0;
+      uint64_t key_offset = 0;
+      // write the blob to the blob log.
+      s = new_writer->AddRecord(record.Key(), record.Blob(),
+        &key_offset, &blob_offset, record.GetTTL());
+
+      new_writer->AddRecordFooter(record.GetSN());
+      newfile->blob_count_++;
+      newfile->file_size_ += blob_log::BlobLogRecord::kHeaderSize +
+        record.Key().size() + record.Blob().size() +
+        blob_log::BlobLogRecord::kFooterSize;
+
+      handle.set_offset(blob_offset);
+      handle.EncodeTo(&index_entry);
+
+      txn->Put(cfh, record.Key(), index_entry);
```

Comment:
If I understand correctly, we are putting all the keys of the file to one single transaction. It makes the transaction more likely to conflict. Can we just do it in key-by-key basis?

When we copy out or delete one key, create one transaction and commit. If there is a conflict, retry.